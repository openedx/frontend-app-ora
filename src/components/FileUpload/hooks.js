import { useState, useReducer, useCallback } from 'react';

export const useUploadConfirmModalHooks = ({
  files, closeHandler, uploadHandler,
}) => {
  const [errors, setErrors] = useState([]);

  const confirmUploadClickHandler = () => {
    const errorList = files.map((file) => (!file.description));
    setErrors(errorList);
    if (errorList.some((error) => error)) {
      return;
    }
    uploadHandler();
  };

  const exitHandler = () => {
    setErrors([]);
    closeHandler();
  };

  // Modifying pointer of file object. This is not a good practice.
  // eslint-disable-next-line no-param-reassign, no-return-assign
  const onFileDescriptionChange = (file) => (event) => file.description = event.target.value;

  return {
    errors,
    confirmUploadClickHandler,
    exitHandler,
    onFileDescriptionChange,
  };
};

export const useFileUploadHooks = ({
  onFileUploaded,
}) => {
  const [uploadState, dispatchUploadState] = useReducer(
    (state, payload) => ({ ...state, ...payload }),
    {
      onProcessUploadArgs: {},
      openModal: false,
    },
  );

  const confirmUpload = useCallback(async () => {
    dispatchUploadState({ openModal: false });
    await onFileUploaded(uploadState.onProcessUploadArgs);
    dispatchUploadState({ onProcessUploadArgs: {} });
  }, [uploadState, onFileUploaded]);

  const closeUploadModal = useCallback(() => {
    dispatchUploadState({ openModal: false, onProcessUploadArgs: {} });
  }, []);

  const onProcessUpload = useCallback(({ fileData, handleError, requestConfig }) => {
    dispatchUploadState({
      onProcessUploadArgs: { fileData, handleError, requestConfig },
      openModal: true,
    });
  }, []);

  return {
    uploadState,
    confirmUpload,
    closeUploadModal,
    onProcessUpload,
  };
};

export default {
  useUploadConfirmModalHooks,
  useFileUploadHooks,
};
