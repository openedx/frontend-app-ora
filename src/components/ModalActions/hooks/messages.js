import { defineMessages } from '@edx/frontend-platform/i18n';

const messages = defineMessages({
  finishLater: {
    id: 'ora-mfe.ModalActions.simpleAction.finishLater',
    defaultMessage: 'Save for later',
    description: 'Save for later (close) button text',
  },
  submitResponse: {
    id: 'ora-mfe.ModalActions.submitResponse',
    defaultMessage: 'Submit response',
    description: 'Submit button text',
  },
  submittingResponse: {
    id: 'ora-mfe.ModalActions.submittingResponse',
    defaultMessage: 'Submitting response',
    description: 'Submit button text while submitting',
  },
  responseSubmitted: {
    id: 'ora-mfe.ModalActions.responseSubmitted',
    defaultMessage: 'Response submitted',
    description: 'Submit button text after successful submission',
  },
  savingResponse: {
    id: 'ora-mfe.ModalActions.savingResponse',
    defaultMessage: 'Saving response',
    description: 'Save for later button text while saving',
  },
});

export default messages;
