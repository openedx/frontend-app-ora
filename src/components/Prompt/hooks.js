import { useState } from 'react';

const usePromptHooks = () => {
  const [open, setOpen] = useState(true);

  const toggleOpen = () => setOpen(!open);

  return {
    open,
    toggleOpen,
  };
};

export default usePromptHooks;
