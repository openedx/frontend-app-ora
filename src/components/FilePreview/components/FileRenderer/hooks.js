import { useState } from 'react';
import { errorStatuses, errorMessages, renderers } from '../constants';
import { getFileType } from '../utils';
import messages from './messages';

/**
 * component hooks
 */
export const useRenderData = ({
  file,
  formatMessage,
}) => {
  const [errorStatus, setErrorStatus] = useState(null);
  const [isLoading, setIsLoading] = useState(true);

  const setState = (newState) => {
    setErrorStatus(newState.errorStatus);
    setIsLoading(newState.isLoading);
  };

  const stopLoading = (status = null) => setState({ isLoading: false, errorStatus: status });

  const errorMessage = (
    errorMessages[errorStatus] || errorMessages[errorStatuses.serverError]
  );
  const errorAction = {
    id: 'retry',
    onClick: () => setState({ errorStatus: null, isLoading: true }),
    message: messages.retryButton,
  };
  const error = {
    headerMessage: errorMessage,
    children: formatMessage(errorMessage),
    actions: [errorAction],
  };

  const Renderer = renderers[getFileType(file.fileName)];
  const rendererProps = {
    fileName: file.fileName,
    url: file.fileUrl,
    onError: stopLoading,
    onSuccess: () => stopLoading(),
  };

  return {
    errorStatus,
    isLoading,
    error,
    Renderer,
    rendererProps,
  };
};
