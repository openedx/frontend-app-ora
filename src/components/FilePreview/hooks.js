import { useKeyedState, StrictDict } from '@edx/react-unit-test-utils';

import {
  PDFRenderer,
  ImageRenderer,
  TXTRenderer,
} from 'components/FilePreview/BaseRenderers';

import messages from './messages';

export const ErrorStatuses = StrictDict({
  badRequest: 400,
  unauthorized: 401,
  forbidden: 403,
  notFound: 404,
  conflict: 409,
  serverError: 500,
});

export const FileTypes = StrictDict({
  pdf: 'pdf',
  jpg: 'jpg',
  jpeg: 'jpeg',
  png: 'png',
  bmp: 'bmp',
  txt: 'txt',
  gif: 'gif',
  jfif: 'jfif',
  pjpeg: 'pjpeg',
  pjp: 'pjp',
  svg: 'svg',
});

/**
 * Config data
 */
export const RENDERERS = {
  [FileTypes.pdf]: PDFRenderer,
  [FileTypes.jpg]: ImageRenderer,
  [FileTypes.jpeg]: ImageRenderer,
  [FileTypes.bmp]: ImageRenderer,
  [FileTypes.png]: ImageRenderer,
  [FileTypes.txt]: TXTRenderer,
  [FileTypes.gif]: ImageRenderer,
  [FileTypes.jfif]: ImageRenderer,
  [FileTypes.pjpeg]: ImageRenderer,
  [FileTypes.pjp]: ImageRenderer,
  [FileTypes.svg]: ImageRenderer,
};

export const SUPPORTED_TYPES = Object.keys(RENDERERS);

export const ERROR_STATUSES = {
  [ErrorStatuses.notFound]: messages.fileNotFoundError,
  [ErrorStatuses.serverError]: messages.unknownError,
};

/**
 * Util methods and transforms
 */
export const getFileType = (fileName) => fileName.split('.').pop()?.toLowerCase();
export const isSupported = (file) => SUPPORTED_TYPES.includes(
  getFileType(file.name),
);

export const stateKeys = StrictDict({
  errorStatus: 'errorStatus',
  isLoading: 'isLoading',
});

/**
 * component hooks
 */
export const renderHooks = ({
  file,
  formatMessage,
}) => {
  const [errorStatus, setErrorStatus] = useKeyedState(stateKeys.errorStatus, null);
  const [isLoading, setIsLoading] = useKeyedState(stateKeys.isLoading, true);

  const setState = (newState) => {
    setErrorStatus(newState.errorStatus);
    setIsLoading(newState.isLoading);
  };

  const stopLoading = (status = null) => setState({ isLoading: false, errorStatus: status });

  const errorMessage = (
    ERROR_STATUSES[errorStatus] || ERROR_STATUSES[ErrorStatuses.serverError]
  );
  const errorAction = {
    id: 'retry',
    onClick: () => setState({ errorStatus: null, isLoading: true }),
    message: messages.retryButton,
  };
  const error = {
    headerMessage: errorMessage,
    children: formatMessage(errorMessage),
    actions: [errorAction],
  };

  const Renderer = RENDERERS[getFileType(file.name)];
  const rendererProps = {
    fileName: file.name,
    url: file.downloadUrl,
    onError: stopLoading,
    onSuccess: () => stopLoading(),
  };

  return {
    errorStatus,
    isLoading,
    error,
    Renderer,
    rendererProps,
  };
};
