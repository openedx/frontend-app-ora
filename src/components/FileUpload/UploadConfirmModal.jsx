import React from 'react';
import PropTypes from 'prop-types';

import {
  Form, FormLabel, ModalDialog, Button, ActionRow,
} from '@edx/paragon';
import { useIntl } from '@edx/frontend-platform/i18n';
import messages from './messages';
import { useUploadConfirmModalHooks } from './hooks';

const UploadConfirmModal = ({
  open, files, closeHandler, uploadHandler,
}) => {
  const { formatMessage } = useIntl();

  const {
    errors, exitHandler, confirmUploadClickHandler, onFileDescriptionChange,
  } = useUploadConfirmModalHooks({
    files,
    closeHandler,
    uploadHandler,
  });

  return (
    <ModalDialog
      isOpen={open}
      title={formatMessage(messages.uploadFileModalTitle)}
      hasCloseButton={false}
      onClose={exitHandler}
    >
      <ModalDialog.Header>
        <ModalDialog.Title>
          {formatMessage(messages.uploadFileModalTitle)}
        </ModalDialog.Title>
      </ModalDialog.Header>

      <ModalDialog.Body>
        <div>
          {files.map((file, i) => (
            // eslint-disable-next-line react/no-array-index-key
            <Form.Group key={i}>
              <FormLabel>
                <strong>
                  {formatMessage(messages.uploadFileDescriptionFieldLabel)}
                </strong>
                <span className="file-name-ellipsis">{file.name}</span>
              </FormLabel>
              <Form.Control
                isInvalid={errors[i]}
                name={`file-${i}-description`}
                onChange={onFileDescriptionChange(file)}
              />
              {errors[i] && (
                <Form.Control.Feedback type="invalid">
                  {errors[i] && formatMessage(messages.fileDescriptionMissingError)}
                </Form.Control.Feedback>
              )}
            </Form.Group>
          ))}
        </div>
      </ModalDialog.Body>
      <ModalDialog.Footer>
        <ActionRow>
          <ModalDialog.CloseButton variant="tertiary" onClick={exitHandler}>
            {formatMessage(messages.cancelUploadFileButton)}
          </ModalDialog.CloseButton>
          <Button variant="primary" onClick={confirmUploadClickHandler}>
            {formatMessage(messages.confirmUploadFileButton)}
          </Button>
        </ActionRow>
      </ModalDialog.Footer>
    </ModalDialog>
  );
};

UploadConfirmModal.defaultProps = {
  open: false,
  files: [],
  closeHandler: () => {},
  uploadHandler: () => {},
};
UploadConfirmModal.propTypes = {
  open: PropTypes.bool,
  files: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.string,
      description: PropTypes.string,
    }),
  ),
  closeHandler: PropTypes.func,
  uploadHandler: PropTypes.func,
};

export default UploadConfirmModal;
