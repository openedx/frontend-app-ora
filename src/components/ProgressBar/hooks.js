import { useParams } from 'react-router-dom';
import { useIsEmbedded, useViewStep } from 'hooks';
import {
  useEffectiveGrade,
  useGlobalState,
} from 'data/services/lms/hooks/selectors';
import {
  stepRoutes,
  stepStates,
} from 'data/services/lms/constants';

export const useProgressStepData = ({ step, canRevisit = false }) => {
  const { xblockId, courseId } = useParams();
  const isEmbedded = useIsEmbedded();
  const viewStep = useViewStep();
  const { stepState } = useGlobalState({ step });

  const href = `/${stepRoutes[step]}${isEmbedded ? '/embedded' : ''}/${courseId}/${xblockId}`;
  const isActive = viewStep === step;
  const isEnabled = (
    isActive
    || (stepState === stepStates.inProgress)
    || (canRevisit && stepState === stepStates.done)
  );
  const myGrade = useEffectiveGrade()?.stepScore;
  return {
    href,
    isEnabled,
    isActive,
    isComplete: stepState === stepStates.done,
    inProgress: stepState === stepStates.inProgress,
    isPastDue: stepState === stepStates.closed,
    myGrade,
    // myGrade: { earned: 8, possible: 10 },
    // isPastDue: step === 'self',
  };
};

export default { useProgressStepData };
