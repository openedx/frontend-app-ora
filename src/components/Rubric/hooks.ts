import { useKeyedState, StrictDict } from '@edx/react-unit-test-utils';
import { usePageData, useRubricConfig } from 'data/services/lms/hooks/selectors';
import { submitRubric } from 'data/services/lms/hooks/actions';
import { RubricHookData } from './types';

export const stateKeys = StrictDict({
  rubric: 'rubric',
  overallFeedback: 'overallFeedback',
});

export const useRubricData = ({ isGrading }): RubricHookData => {
  const data = usePageData();
  const { criteria, feedbackConfig } = useRubricConfig();

  const [rubricData, setRubricData] = useKeyedState(stateKeys.rubric, data.rubric);
  const [overallFeedback, setOverallFeedback] = useKeyedState(
    stateKeys.overallFeedback, data.rubric.overallFeedback
  );
  const submitRubricMutation = submitRubric();

  const onOverallFeedbackChange = (e: React.ChangeEvent<HTMLInputElement>) =>
    setOverallFeedback(e.target.value);

  const onSubmit = () => {
    submitRubricMutation.mutate({
      overallFeedback,
      optionsSelected: rubricData.optionsSelected,
      criterionFeedback: rubricData.criterionFeedback,
    } as any);
  };

  return {
    rubricData,
    setRubricData,
    criteria: criteria.map((criterion) => ({
      ...criterion,
      optionsValue: rubricData.optionsSelected[criterion.name],
      optionsIsInvalid: false,
      optionsOnChange: (e: React.ChangeEvent<HTMLInputElement>) => {
        setRubricData({
          ...rubricData,
          optionsSelected: {
            ...rubricData.optionsSelected,
            [criterion.name]: e.target.value,
          },
        });
      },

      feedbackValue: rubricData.criterionFeedback[criterion.name],
      feedbackIsInvalid: false,
      feedbackOnChange: (e: React.ChangeEvent<HTMLInputElement>) => {
        setRubricData({
          ...rubricData,
          criterionFeedback: {
            ...rubricData.criterionFeedback,
            [criterion.name]: e.target.value,
          },
        });
      },
    })),
    onSubmit,
    submitStatus: submitRubricMutation.status,

    // overall feedback
    overallFeedback,
    onOverallFeedbackChange,
    overallFeedbackDisabled: !isGrading,
    overallFeedbackIsInvalid: false,
    overallFeedbackPrompt: feedbackConfig.defaultText,
  };
};

export default useRubricData;
