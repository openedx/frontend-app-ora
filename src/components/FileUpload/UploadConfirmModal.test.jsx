import { shallow } from '@edx/react-unit-test-utils';
import UploadConfirmModal from './UploadConfirmModal';

import { useUploadConfirmModalHooks } from './hooks';

jest.mock('./hooks', () => ({
  useUploadConfirmModalHooks: jest.fn(),
}));

describe('<UploadConfirmModal />', () => {
  const props = {
    open: true,
    files: [],
    closeHandler: jest.fn().mockName('closeHandler'),
    uploadHandler: jest.fn().mockName('uploadHandler'),
  };

  const mockHooks = (overrides) => {
    useUploadConfirmModalHooks.mockReturnValueOnce({
      errors: [],
      exitHandler: jest.fn().mockName('exitHandler'),
      confirmUploadClickHandler: jest.fn().mockName('confirmUploadClickHandler'),
      onFileDescriptionChange: () => jest.fn().mockName('onFileDescriptionChange'),
      ...overrides,
    });
  };
  describe('renders', () => {
    test('no files', () => {
      mockHooks();
      const wrapper = shallow(<UploadConfirmModal {...props} />);
      expect(wrapper.snapshot).toMatchSnapshot();

      expect(wrapper.instance.findByType('Form.Group').length).toBe(0);
    });

    test('multiple files', () => {
      mockHooks(
        { errors: new Array(2) },
      );
      const wrapper = shallow(<UploadConfirmModal {...props} files={[{ name: 'file1' }, { name: 'file2' }]} />);
      expect(wrapper.snapshot).toMatchSnapshot();

      expect(wrapper.instance.findByType('Form.Group').length).toBe(2);
      expect(wrapper.instance.findByType('Form.Control.Feedback').length).toBe(0);
    });

    test('with errors', () => {
      mockHooks({ errors: [true, false] });
      const wrapper = shallow(<UploadConfirmModal {...props} files={[{ name: 'file1' }, { name: 'file2' }]} />);
      // wrapper.setState({ errors: [true, false] });
      expect(wrapper.snapshot).toMatchSnapshot();

      expect(wrapper.instance.findByType('Form.Group').length).toBe(2);
      expect(wrapper.instance.findByType('Form.Control.Feedback').length).toBe(1);
    });
  });
});
