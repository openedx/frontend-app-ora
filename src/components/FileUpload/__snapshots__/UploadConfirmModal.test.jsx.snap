// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UploadConfirmModal /> renders multiple files 1`] = `
<ModalDialog
  hasCloseButton={false}
  isOpen={true}
  onClose={[MockFunction exitHandler]}
  title="Add a text description to your file"
>
  <ModalDialog.Header>
    <ModalDialog.Title>
      Add a text description to your file
    </ModalDialog.Title>
  </ModalDialog.Header>
  <ModalDialog.Body>
    <div>
      <Form.Group
        key="0"
      >
        <FormLabel>
          <strong>
            Description for: 
          </strong>
          <span
            className="file-name-ellipsis"
          >
            file1
          </span>
        </FormLabel>
        <Form.Control
          name="file-0-description"
          onChange={[MockFunction onFileDescriptionChange]}
        />
      </Form.Group>
      <Form.Group
        key="1"
      >
        <FormLabel>
          <strong>
            Description for: 
          </strong>
          <span
            className="file-name-ellipsis"
          >
            file2
          </span>
        </FormLabel>
        <Form.Control
          name="file-1-description"
          onChange={[MockFunction onFileDescriptionChange]}
        />
      </Form.Group>
    </div>
  </ModalDialog.Body>
  <ModalDialog.Footer>
    <ActionRow>
      <ModalDialog.CloseButton
        onClick={[MockFunction exitHandler]}
        variant="tertiary"
      >
        Cancel upload
      </ModalDialog.CloseButton>
      <Button
        onClick={[MockFunction confirmUploadClickHandler]}
        variant="primary"
      >
        Upload files
      </Button>
    </ActionRow>
  </ModalDialog.Footer>
</ModalDialog>
`;

exports[`<UploadConfirmModal /> renders no files 1`] = `
<ModalDialog
  hasCloseButton={false}
  isOpen={true}
  onClose={[MockFunction exitHandler]}
  title="Add a text description to your file"
>
  <ModalDialog.Header>
    <ModalDialog.Title>
      Add a text description to your file
    </ModalDialog.Title>
  </ModalDialog.Header>
  <ModalDialog.Body>
    <div />
  </ModalDialog.Body>
  <ModalDialog.Footer>
    <ActionRow>
      <ModalDialog.CloseButton
        onClick={[MockFunction exitHandler]}
        variant="tertiary"
      >
        Cancel upload
      </ModalDialog.CloseButton>
      <Button
        onClick={[MockFunction confirmUploadClickHandler]}
        variant="primary"
      >
        Upload files
      </Button>
    </ActionRow>
  </ModalDialog.Footer>
</ModalDialog>
`;

exports[`<UploadConfirmModal /> renders with errors 1`] = `
<ModalDialog
  hasCloseButton={false}
  isOpen={true}
  onClose={[MockFunction exitHandler]}
  title="Add a text description to your file"
>
  <ModalDialog.Header>
    <ModalDialog.Title>
      Add a text description to your file
    </ModalDialog.Title>
  </ModalDialog.Header>
  <ModalDialog.Body>
    <div>
      <Form.Group
        key="0"
      >
        <FormLabel>
          <strong>
            Description for: 
          </strong>
          <span
            className="file-name-ellipsis"
          >
            file1
          </span>
        </FormLabel>
        <Form.Control
          isInvalid={true}
          name="file-0-description"
          onChange={[MockFunction onFileDescriptionChange]}
        />
        <Form.Control.Feedback
          type="invalid"
        >
          Please enter a file description
        </Form.Control.Feedback>
      </Form.Group>
      <Form.Group
        key="1"
      >
        <FormLabel>
          <strong>
            Description for: 
          </strong>
          <span
            className="file-name-ellipsis"
          >
            file2
          </span>
        </FormLabel>
        <Form.Control
          isInvalid={false}
          name="file-1-description"
          onChange={[MockFunction onFileDescriptionChange]}
        />
      </Form.Group>
    </div>
  </ModalDialog.Body>
  <ModalDialog.Footer>
    <ActionRow>
      <ModalDialog.CloseButton
        onClick={[MockFunction exitHandler]}
        variant="tertiary"
      >
        Cancel upload
      </ModalDialog.CloseButton>
      <Button
        onClick={[MockFunction confirmUploadClickHandler]}
        variant="primary"
      >
        Upload files
      </Button>
    </ActionRow>
  </ModalDialog.Footer>
</ModalDialog>
`;
